modules/prebid/optimization/rulesengine/schema_functions.go

```diff
@@ -0,0 +1,475 @@
+package rulesengine
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"github.com/prebid/openrtb/v20/adcom1"
+	"github.com/prebid/prebid-server/v3/openrtb_ext"
+	"github.com/prebid/prebid-server/v3/util/jsonutil"
+	"github.com/prebid/prebid-server/v3/util/ptrutil"
+	"math/rand"
+	"slices"
+)
+
+type SchemaFunction interface {
+	Call(wrapper *openrtb_ext.RequestWrapper) (string, error)
+}
+
+func NewSchemaFunctionFactory(name string, params json.RawMessage) (SchemaFunction, error) {
+	switch name {
+	case "deviceCountry":
+		return NewDeviceCountry(params)
+	case "dataCenters":
+		return NewDataCenter(params)
+	case "channel":
+		return NewChannel()
+	case "eidAvailable":
+		return NewAidAvailable(params)
+	case "userFpdAvailable":
+		return NewUserFpdAvailable()
+	case "fpdAvail":
+		return NewFpdAvail()
+	case "gppSid":
+		return NewGppSid(params)
+	case "tcfInScope":
+		return NewTcfInScope()
+	case "percent":
+		return NewPercent(params)
+	case "prebidKey":
+		return NewPrebidKey(params)
+	case "domain":
+		return NewDomain(params)
+	case "bundle":
+		return NewBundle(params)
+	case "deviceType":
+		return NewDeviceType(params)
+
+	default:
+		return nil, fmt.Errorf("Schema function %s was not created", name)
+	}
+}
+
+// ------------deviceCountry------------------
+type deviceCountry struct {
+	CountryCodes []string
+}
+
+func NewDeviceCountry(params json.RawMessage) (SchemaFunction, error) {
+	var devCountry []string
```

Comment:
All of the schema functions take `params` and unmarshal them into an array of strings `[]string` but, in Brian's spec doc comment it is established that args is an array of arrays of strings. Should we modify all of these?
- `NewDeviceCountry(params)`
- `NewDataCenter(params)`
- `NewChannel()`
- `NewAidAvailable(params)`
- `NewUserFpdAvailable()`
- `NewFpdAvail()`
- `NewGppSid(params)`
- `NewTcfInScope()`
- `NewPercent(params)`
- `NewPrebidKey(params)`
- `NewDomain(params)`
- `NewBundle(params)`
- `NewDeviceType(params)`

Reply:
Params can have a different type for different schema functions. This is why we have this kind of a structure we unmarshal json config into:
```golang
type Schema struct {
	Func string          `json:"function,omitempty"`
	Args json.RawMessage `json:"args,omitempty"`
}
```
Here are some schema function descriptions from the doc: 
Function           | Args 
deviceCountry | null OR array of strings 
channel             | null
gppSid              | null OR array of ints
percent             | int
prebidKey         | string
bidPrice            | [OPERATOR, VALUE, CURRENCY]


This is the reason why we pass json.RawMessage as a parameter. Every schema function knows what type to expect. 

Do I miss something?